File: app/src/main/java/com/arcinteractive/spaces/data/model/Ping.kt
package com.arcinteractive.spaces.data.model

import java.text.DateFormat
import java.util.Date

data class Ping(
    val id: String,
    val participantIds: List<String>,
    val participantNames: List<String>,
    val participantEmojis: List<String>,
    val lastMessageAt: Date?,
    val lastMessagePreviewType: String?,
    val createdAt: Date?,
    val updatedAt: Date?,
    val unreadCount: Int
) {
    fun otherParticipant(currentUserId: String?): PingParticipant? {
        if (currentUserId == null) return null
        val index = participantIds.indexOfFirst { it != currentUserId }
        if (index < 0) return null
        return PingParticipant(
            id = participantIds[index],
            displayName = participantNames.getOrNull(index) ?: "User",
            emojiAvatar = participantEmojis.getOrNull(index) ?: "🙂"
        )
    }

    fun title(currentUserId: String?): String {
        return otherParticipant(currentUserId)?.displayName ?: participantNames.firstOrNull() ?: "Ping"
    }

    fun emoji(currentUserId: String?): String {
        return otherParticipant(currentUserId)?.emojiAvatar ?: participantEmojis.firstOrNull() ?: "\uD83D\uDCAC"
    }

    val timestampText: String
        get() = dateFormatter.format(lastMessageAt ?: updatedAt ?: createdAt ?: Date())

    val lastMessagePreviewText: String
        get() = when (lastMessagePreviewType) {
            "video" -> "Video"
            "image" -> "Photo"
            "file" -> "File"
            "text" -> "Message"
            else -> "No messages yet"
        }
}

data class PingParticipant(
    val id: String,
    val displayName: String,
    val emojiAvatar: String
)

private val dateFormatter: DateFormat = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.SHORT)

File: iOS/Spaces V1.5/Models/Ping.swift
import Foundation

struct PingParticipant: Identifiable, Hashable {
    let id: String
    let displayName: String
    let emojiAvatar: String
}

struct Ping: Identifiable, Hashable {
    let id: String
    let participantIds: [String]
    let participantNames: [String]
    let participantEmojis: [String]
    let lastMessageAt: Date?
    let lastMessagePreviewType: String?
    let createdAt: Date?
    let updatedAt: Date?
    let unreadCount: Int

    func otherParticipant(for currentUserID: String?) -> PingParticipant? {
        guard let currentUserID else { return nil }
        guard let index = participantIds.firstIndex(where: { $0 != currentUserID }) else { return nil }
        return PingParticipant(
            id: participantIds[index],
            displayName: participantNames[safe: index] ?? "User",
            emojiAvatar: participantEmojis[safe: index] ?? "🙂"
        )
    }

    func title(for currentUserID: String?) -> String {
        otherParticipant(for: currentUserID)?.displayName ?? participantNames.first ?? "Ping"
    }

    func emoji(for currentUserID: String?) -> String {
        otherParticipant(for: currentUserID)?.emojiAvatar ?? participantEmojis.first ?? "💬"
    }

    var timestampText: String {
        guard let referenceDate = lastMessageAt ?? updatedAt ?? createdAt else { return "" }
        return Self.relativeTimestampFormatter.localizedString(for: referenceDate, relativeTo: Date())
    }

    var lastMessagePreviewText: String {
        switch lastMessagePreviewType {
        case MessageType.video.rawValue:
            return "Video"
        case MessageType.image.rawValue:
            return "Photo"
        case MessageType.file.rawValue:
            return "File"
        case MessageType.text.rawValue:
            return "Message"
        default:
            return "No messages yet"
        }
    }

    private static let relativeTimestampFormatter: RelativeDateTimeFormatter = {
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .short
        return formatter
    }()
}

private extension Array {
    subscript(safe index: Int) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}
